### PR TITLE
docs: extend extension documentation

### PR DESCRIPTION
**Pull request by jvanstraten**
- Coin the term "core extensions" and describe why they (and extensions in general) exist.
 - Document guidelines from #251 and #307.
 - Document common options in core extensions (#254).
 - Better integrate the function documentation generator output into the website.

~~This depends on #309 (its commit as initially proposed is included in the history here) so I've marked this as draft for now. There's also a hidden conflict with #302, since this documents the option that a change is being proposed for there. I'll rebase/update when both of those are merged.~~

### CODE REVIEW

**Review by jacques-n — COMMENTED** [review]
Really like this update. Thanks @jvanstraten !

Will do a second pass once out of draft form.

**Review by jacques-n — COMMENTED** [review]
Mostly looks good, a couple of comments. Thanks @jvanstraten

File: site/docs/extensions/core_extensions/guidelines.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    - More information and examples: [#251](https://github.com/substrait-io/substrait/issues/251)
+
+ - Be pedantic when describing functionality. The corner cases that rarely come up in practice are exactly the places where different implementations are likely to differ, so for a plan to be implementation-agnostic, these are exactly the things that need to be specified exhaustively. For especially pedantic things, an optional enumeration argument may be suitable; this allows a producer to explicitly indicate that the consumer can pick the behavior.
+    - Example: the verbosity of the description of [regex_match_substring](https://github.com/substrait-io/substrait/blob/fbe5e0949b863334d02b5ad9ecac55ec8fc4debb/extensions/functions_string.yaml#L79-L139).
+    - Example: the floating point rounding option defined [here](common_options.md).
+
+ - The core extensions should generally not be defining type classes. If you believe a type class that isn't currently in the specification is important enough to include, it probably makes more sense to simply add it to the built-in types, or otherwise should be a third-party extension.
```
[the remark this answers is not in the document]
Comment (jvanstraten) [reply site/docs/extensions/core_extensions/guidelines.md:28]: Hm. The way I figured things would go is that data representation projects like Arrow would publish and maintain extension types of their own, that everyone else can then use if they use that data format under the hood. I'm not sure Substrait should necessarily be involved with that.

File: site/docs/extensions/core_extensions/guidelines.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    - Example: [#287 (comment)](https://github.com/substrait-io/substrait/pull/287#discussion_r942705485)
+
+ - Avoid adding functions that express behaviors that are already expressible with [specialized record expressions](../../expressions/specialized_record_expressions/).
+    - Example: the `if_else` function originally proposed in [#291](https://github.com/substrait-io/substrait/issues/291)
+
+ - Prefer adding new options to existing functions instead of adding new functions.
+    - Example: [#289](https://github.com/substrait-io/substrait/issues/289)
+    - Example: [#295](https://github.com/substrait-io/substrait/issues/295)
+
+ - Aim for syntactic and semantic consistency with widely used SQL dialects, especially PostgreSQL.
```
[the remark this answers is not in the document]
Comment (jvanstraten) [reply site/docs/extensions/core_extensions/guidelines.md:15]: I copypasted this from #307 and feel a bit out of my league here. @ianmcook, any comment? If not I'll just remove it.

File: site/docs/extensions/core_extensions/guidelines.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    - Example: [#285 (comment)](https://github.com/substrait-io/substrait/pull/285#discussion_r944542030)
+
+ - Generalize the function as much as possible, to reduce the odds that we'll need to update it later.
+    - Example: for a function like `add`, consider making it variadic rather than only accepting two arguments.
+
+ - Be consistent when it comes to argument types. It is preferable to define a function that accepts and returns one type class over a function that promotes from one type class or another or accepts a mixture of type classes. This aims to prevent an explosion of function implementations.
+    - More information and examples: [#251](https://github.com/substrait-io/substrait/issues/251)
+
+ - Be pedantic when describing functionality. The corner cases that rarely come up in practice are exactly the places where different implementations are likely to differ, so for a plan to be implementation-agnostic, these are exactly the things that need to be specified exhaustively. For especially pedantic things, an optional enumeration argument may be suitable; this allows a producer to explicitly indicate that the consumer can pick the behavior.
```
[the remark this answers is not in the document]
Comment (jvanstraten) [reply site/docs/extensions/core_extensions/guidelines.md:24]: IMO "pedantic" is not a bad thing for a specification, but, sure. "Precise" as a drop-in replacement maybe?

File: site/docs/extensions/core_extensions/guidelines.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jvanstraten) [reply site/docs/extensions/core_extensions/guidelines.md]: Removed statement in 22fd8c3.

File: site/docs/extensions/core_extensions/guidelines.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jvanstraten) [reply site/docs/extensions/core_extensions/guidelines.md]: Removed it in 22fd8c3

**Comment by jacques-n** [comment]
No progress has been made in more than six months. Closing without prejudice.

**Pull request closed without merging**